server/src/test/java/org/elasticsearch/cluster/health/NoDataNodesHealthTests.java

```diff
@@ -0,0 +1,153 @@
+
+import java.util.Collections;
+
+import static org.elasticsearch.cluster.routing.ShardRoutingState.INITIALIZING;
+import static org.elasticsearch.cluster.routing.ShardRoutingState.STARTED;
+
+/**
+ * The test case checks for the scenario when there is no data node in the cluster and only
+ * master is active. At this moment when index creation is tried, the cluster health status should
+ * change to RED
+ */
+public class NoDataNodesHealthTests extends ESAllocationTestCase {
+    /**
+     * This method specifically creates a cluster with no data nodes
+     * and a single master node
+     */
+    private ClusterState setUpClusterWithNoDataNodes() {
+
+        DiscoveryNodes node = DiscoveryNodes.builder().add(newNode("node_m", Collections.singleton(DiscoveryNode.Role.MASTER))).build();
+        MetaData metaData = MetaData.builder()
+                .put(IndexMetaData.builder("TestIndex")
+                        .settings(settings(Version.CURRENT))
+                        .numberOfShards(randomIntBetween(1, 3))
+                        .numberOfReplicas(randomIntBetween(0, 2)))
+                .build();
+        RoutingTable routingTable = RoutingTable.builder().addAsNew(metaData.index("TestIndex")).build();
+        ClusterState state = ClusterState.builder(new ClusterName("test_cluster"))
+                .nodes(node)
+                .metaData(metaData)
+                .routingTable(routingTable)
+                .build();
+        MockAllocationService service = createAllocationService();
+        state = service.reroute(state, "reroute");
+        return state;
+    }
+
+    public void testClusterHealthWithNoDataNodes() {
+        ClusterState state = setUpClusterWithNoDataNodes();
+        int dataNodes = state.nodes().getDataNodes().size();
+        int masterNodes = state.nodes().getMasterNodes().size();
+        assertTrue(dataNodes == 0);
+        assertTrue(masterNodes > 0);
+        ClusterHealthStatus clusterHealthStatus = new ClusterStateHealth(state).getStatus();
+        assertEquals(ClusterHealthStatus.RED, clusterHealthStatus);
+    }
+
+    /**
+     * The method test for scenario where we have a cluster with indices and data nodes
+     * and then all data nodes gets terminated now new index is created then
+     * all indices and cluster health should be red, but last allocation attempts of new index shards v/s
+     * old index shards should be different.
+     */
+    public void testAllocationStatusForTerminatedNodes() {
+        //creates one master and two data nodes
+        DiscoveryNodes.Builder nodeBuilder = DiscoveryNodes.builder().add(newNode("node_m", Collections.singleton(DiscoveryNode.Role.MASTER)))
+                .add(newNode("node_d1", Collections.singleton(DiscoveryNode.Role.DATA)))
+                .add(newNode("node_d2", Collections.singleton(DiscoveryNode.Role.DATA)));
+        MetaData metaData = MetaData.builder()
+                .put(IndexMetaData.builder("TestIndex")
```

Comment:
Elsewhere in the test suite indices normally have all-lowercase names:

```suggestion
                .put(IndexMetaData.builder("test")
```